### PR TITLE
Workflow changes for GLM Flash Extent Density Assimilation

### PR DESCRIPTION
**Pull request by AmandaBack-NOAA**
## DESCRIPTION OF CHANGES: 
Added scripts for observation preprocessing
Made changes to workflow generation to add preprocessing when DO_GLM_FED_DA="TRUE" (it is "FALSE" by default)
Most changes are in place for assimilation - run_analysis script and anavinfo changes will come separately

## TESTS CONDUCTED: 
Confirmed that when DO_GLM_FED_DA="FALSE" both parts of coupled workflow produce identical results to workflow before the PR
Confirmed that when DO_GLM_FED_DA="TRUE" the preprocessor runs successfully and no other parts of the workflow are disrupted
Science changes were evaluated for the previous PRs by David Dowell and Hongli Wang that updated GSI and EnKF code
- [ ] WCOSS2
- [ x] Hera
- [ ] Orion
- [ ] Hercules
- [ x] Jet

## CONTRIBUTORS (optional):
Ming Hu, David Dowell, Hongli Wang

### CODE REVIEW

**Comment by chan-hoo** [comment]
@AmandaBack-NOAA, please open an issue for this PR and add its number to the description of this PR.

**Comment by AmandaBack-NOAA** [comment]
@chan-hoo Done.

File: fix/gsi/glmwest_full_2km_lat_lon.nc
Comment (guoqing-noaa) [inline fix/gsi/glmwest_full_2km_lat_lon.nc:1]: The 3 glm* fix files are available on Jet/Hera. I also sync'ed them to Orion/Hercules.
@ShunLiu-NOAA could you check whether they are available on wcoss2? If not, they can be copied from one of the above platforms.

Comment (MatthewPyle-NOAA) [reply fix/gsi/glmwest_full_2km_lat_lon.nc:1]: @guoqing-noaa I'll work to get them on wcoss2 if needed.

Comment (MatthewPyle-NOAA) [reply fix/gsi/glmwest_full_2km_lat_lon.nc:1]: @MatthewPyle-NOAA - got them onto cactus (in `/lfs/h2/emc/lam/noscrub/emc.lam/FIX_RRFS/gsi`), but need to remember to get them onto dogwood once that system returns.

Comment (guoqing-noaa) [reply fix/gsi/glmwest_full_2km_lat_lon.nc:1]: Thanks, @MatthewPyle-NOAA !

**Comment by daviddowellNOAA** [comment]
@AmandaBack-NOAA Thanks for preparing this important PR.  I'll report here on my testing on Jet with a case from August 2023.

First, like you, I ran your workflow with DO_GLM_FED_DA = FALSE.  I agree that it produces identical results to the unmodified rrfs-workflow for all EnVar and EnKF data-assimilation tasks with conventional and radar-reflectivity observations.

Second, I tried DO_GLM_FED_DA = TRUE for the ensemble workflow.  The FED observation processing was successful.

Some changes are needed to exrrfs_run_analysis.sh to allow the GSI observer to use FED observations.  Take a look at my version of the script on Jet:
/mnt/lfs1/BMC/wrfruc/HRRRE/code/rrfs-workflow.amanda/scripts/exrrfs_run_analysis.sh
The changes are (1) link to the FED observation file, (2) set diag_fed to true when DO_GLM_FED_DA is true, and (3) add conv_fed to the list of diag files.

With these changes to exrrfs_run_analysis.sh, I was able to get a successful run of the later ENKF_RADAR_REF task, assimilating radar-reflectivity and FED observations simultaneously into the ensemble.

**Comment by chan-hoo** [comment]
@AmandaBack-NOAA @daviddowellNOAA, could you let us know which sample script we can use to test this PR?

**Comment by AmandaBack-NOAA** [comment]
@daviddowellNOAA Thanks. I tested your changes on Hera and added them to my fork/this PR.

@chan-hoo This PR makes updates to the scripts and workflows and does not change any executables. The way to test it is by building and running the workflow, which you can do by checking out my fork that the PR comes from. If you'd like to test with ensemble GLM assimilation turned on, the new config variables to edit are DO_GLMFED_DA and GLMFED_DATA_MODE. I added examples on lines 25-26 of ush/sample_configs/others/config.sh.RRFS_CONUS_3km_ens. That's the only configuration in sample_configs/others in which DO_GLMFED_DA should be changed from its default of "FALSE" to "TRUE" for now. If you use a sample config that is not in that directory for 3km ensemble tests, you can add those two lines to your config file.

File: fix/gsi/glmwest_full_2km_lat_lon.nc
Comment (MatthewPyle-NOAA) [reply fix/gsi/glmwest_full_2km_lat_lon.nc:1]: The glm fixed files now are on the dogwood side of WCOSS as well.

**Comment by chan-hoo** [comment]
@AmandaBack-NOAA, when running the sample configuration, should I run `rocotoboot` for the `prep_cyc_spinup_ensinit_memXXX` tasks on `202202010600` manually?

**Comment by chan-hoo** [comment]
@AmandaBack-NOAA, my test run failed due to missing (restart) files at `06z` like:
```
  cp: cannot stat '/lfs4/HFIP/hfv3gfs/Chan-hoo.Jeon/rrfs_test_da_ens/stmp/2022020106/fcst_fv3lam_spinup/INPUT/fv_core.res.tile1.nc': No such file or directory
```
Is there an option for a cold start in this configuration?

**Comment by AmandaBack-NOAA** [comment]
> @AmandaBack-NOAA, my test run failed due to missing (restart) files at `06z` like:
> 
> ```
>   cp: cannot stat '/lfs4/HFIP/hfv3gfs/Chan-hoo.Jeon/rrfs_test_da_ens/stmp/2022020106/fcst_fv3lam_spinup/INPUT/fv_core.res.tile1.nc': No such file or directory
> ```
> 
> Is there an option for a cold start in this configuration?

@chan-hoo In response to your two questions:
 - It should not be necessary to boot any tasks. It should start by itself when dependencies are met.
 - For that error, it looks like you must have booted a task. Is that right? The error that you see is because you pointed it to an STMP directory "/lfs4/HFIP/hfv3gfs/Chan-hoo.Jeon/rrfs_test_da_ens/stmp" in which to find the *control member* output, but if you are not running a control member workflow, that will not be found. You will need to run the control member through its first 4 cycles to get the first cycle of the ensemble workflow to run. This is standard to the ensemble workflow and not a change that we made for the PR.

**Comment by chan-hoo** [comment]
@AmandaBack-NOAA, in your comment `You will need to run the control member through its first 4 cycles to get the first cycle of the ensemble workflow to run.`, this is what I want to do. How can I run the first 4 cycles? Can you let me know how to change the configuration for this 4 cycles?

**Comment by AmandaBack-NOAA** [comment]
@chan-hoo Right now you are running the ensemble workflow. You need the coupled control member workflow. When you have a control member workflow that you are running, none of the tasks will end with _mem00XX like the prep_cyc_mem00XX that you mention.

When you have a control member workflow, you will need to boot the first prep_cyc_spinup task at 3Z, but don't boot it until the foregoing tasks get_extrn_ics, make_ics, get_extrn_lbcs, and make_lbcs have run.

**Comment by chan-hoo** [comment]
@AmandaBack-NOAA, thanks for sharing the doc!!! Does this mean that we are not able to run an ensemble case with any sample scripts alone?

**Comment by AmandaBack-NOAA** [comment]
@chan-hoo If someone else already has those first few control member cycles run and has not cleaned their stmp directories, you can point your ensemble workflow to their stmp to read. I have the first cycles for 20230610 on Hera that you can use. @daviddowellNOAA probably has other hours on Jet, and if not, I could also run them on Jet for you to use.

**Comment by chan-hoo** [comment]
@AmandaBack-NOAA, thanks for the suggestion! I am trying to add a new engineering test case for an ensemble run. I'll take a look at it first and let you know soon.

**Comment by chan-hoo** [comment]
@AmandaBack-NOAA, can you point me to your stmp directory for `20230610` on Hera?

**Comment by chan-hoo** [comment]
@MatthewPyle-NOAA, I wanted to set up a new engineering test case for an ensemble run through this PR. However, as you can see above, it will take some time. I think @AmandaBack-NOAA and @daviddowellNOAA have tested this change with their data. You don't have to wait for my approval. If this is an urgent task, please go ahead.

**Comment by MatthewPyle-NOAA** [comment]
Thanks for the comment, @chan-hoo.  I'll try to take another look at it today, and will approve/merge if I don't have any more concerns.

**Review by MatthewPyle-NOAA — APPROVED** [review]
Will approve, but did have one small question for @AmandaBack-NOAA.  Looking ahead to running GLM in the real-time system on WCOSS, will it just be a matter of defining appropriate GLMFED_EAST_ROOT and GLMFED_WEST_ROOT for that system?

**Comment by AmandaBack-NOAA** [comment]
@MatthewPyle-NOAA As far as I know, the following changes are needed to your config file before generating the RRFS A ensemble workflow, to perform GLM flash extent data assimilation:

 - update GLMFED_[EAST/WEST]_ROOT
 - include the line DO_GLM_FED_DA="TRUE"
 - include the line GLMFED_DATA_MODE="EMC"
 
If anyone runs tests like this on WCOSS2, please copy logs and, if possible, fedobs.nc to Jet or Hera to share with me. My WCOSS2 access is still pending so I can't check anything there.

**Comment by MatthewPyle-NOAA** [comment]
Thanks @AmandaBack-NOAA for the details.  Will get this one merged.

**Pull request merged**
